subsys/bluetooth/host/att.c

```diff
@@ -605,7 +633,44 @@ static u8_t find_type_cb(const struct bt_gatt_attr *attr, void *user_data)
 	}
 
 	/* Check if data matches */
-	if (read != data->value_len || memcmp(data->value, uuid, read)) {
+	if (read != data->value_len) {
+		/* Use bt_uuid_cmp() to compare UUIDs of different form. */
+		union {
+			struct bt_uuid_16 u16;
+			struct bt_uuid_32 u32;
+			struct bt_uuid_128 u128;
+		} ref_uuid, recvd_uuid;
+		union {
+			u16_t *u16;
+			u32_t *u32;
+			u8_t *u128;
+		} v;
+
+		/* Copy from packet data to struct bt_uuid format */
+		bt_uuid_create(&recvd_uuid.u128.uuid, data->value,
+			       data->value_len);
+		/* Copy from internal list to struct bt_uuid format */
+		v.u128 = uuid;
+		switch (read) {
```

Comment:
@joerchan , the attr->read() function reads from the internal list of characteristics in CPU endianness where bt_uuid_create() reads from the receivied packet which is in little endian format.

Reply:
I understand. Would it perhaps be cleaner to create a different function for that as well? And document when they both can be used?
Also the uuid_create (net_buf version) could also re-use bt_uuid_create by using the req->uuid. 
IMO it would make this cleaner to have one function to create uuid from the received packet.

Reply:
@joerchan , I agree that it would be cleaner, and have also considered to create a separate function, but I stopped because it would be a function that seems to be only related to the attribute read callback. Would we use the functionality at other places, or can I just put it in att.c as a local function?
With regard to the uuid_create(), I am on thin ice. I am a little bit afraid to break code by not doing proper buf->len and buf->data manipulation. But I can give it a try, and then the reviewers can vote.

Reply:
I think we should make them equally available, so perhaps `bt_uuid_create`, and `bt_uuid_create_le` the latter one creating uuid from Little Endian data format.

Regarding uuid_create(), as you pointed out here: https://github.com/zephyrproject-rtos/zephyr/pull/17765#discussion_r307730536 the buf->len is inconsistely moved depending on the UUID type. To me it seems like the reason this is not a problem is because the buf is not used after, otherwise I don't understand how that could not be an issue.
It seems like this is not the intention, and could lead to problems in the future. That is a very unexpected behaviour of the function IMO.

Reply:
@joerchan , I have replaced uuid_create() with my bt_uuid_create() function, and it seems to work (at least for my test case).